Multi image (#1508)
NordicPlayground/fw-nrfconnect-nrf: doc/nrf/ug_multi_image.rst

Multi-image builds
##################

In many cases, the firmware that is programmed to a device consists of not only one application, but several separate images, where one of the images (the *parent image*) requires one or more other images (the *child images*) to be present.

> The parent image then *chain-loads* (or *boots*) the child image, which in turn might boot another child image.

Reviewer: It's the other way around, children boot their parents.

app = parent
bootloader = child

A bootloader boots an application.

Reviewer: Also, currently some children boot their siblings.

Technically, there is nothing preventing a parent from booting a child, it's just not been done yet, and there is no indication it will be done in the future either.

Reviewer 2: These comments looks weird out of context :D

Reviewer 3: Updated to:
The child image then *chain-loads* (or *boots*) the parent image, which in turn might be a child image to another parent image and boot that one.

When to use multiple images
***************************

An *image* (also referred to as executable, program, or elf file) consists of pieces of code and data that are identified by image-unique names recorded in a single symbol table.
The symbol table exists as metadata in a ``.elf`` or ``.exe`` file and is not included when the image is converted to a HEX file for programming.
Instead, the code and data are placed at addresses by a linker.
This linking process is what distinguishes images from object files (which do not require linking).
Therefore, to determine if you have zero, one, or more images, count the number of times the linker runs.

Using multiple images has the following advantages:

* You can run the linker multiple times and partition the final firmware into several regions.
  This partitioning is often useful for bootloaders.
* Since there is a symbol table for each image, the same symbol names can exist multiple times in the final firmware.
  This is useful for bootloader images, because it makes it possible to reuse libraries that the application image uses in the bootloader image.

>   (? is this correct? Original input was "This again is useful for bootloader images, as they often need to have their own fixed copies of the same libraries that the application image is using." ?)

Reviewer: I wouldn't word it like this. reuse might imply the library only exists once in flash, but the whole point is that you might need the same library (same sources and symbol names at least) configured differently in two different places in flash.

Reviewer: Might even be the "same" library, but two different versions, so not the exact same sources per se, but still the same problem of symbol name aliasing issues which require that the libraries be placed in two different images.

Reviewer 3: Updated to: 
This is useful for bootloader images, which might required their own copy of a library that the application uses, but in a different version or configuration.

   Therefore, the nRF9160 samples (the parent image) require the :ref:`secure_partition_manager` (the child image) to be programmed in addition to the actual application.

   See :ref:`zephyr:nrf9160_pca10090` and :ref:`ug_nrf9160` for more information.

MCUboot bootloader
   The MCUboot bootloader establishes a root of trust by verifying the next step in the boot sequence.
   This first-stage bootloader is immutable, which means it must never be updated or deleted.
   However, it allows to update the application, and therefore MCUboot and the application must be located in different images.
   In this scenario, the application is the parent image and MCUboot is the child image.

   See :ref:`about_mcuboot` for more information.
   The MCUboot bootloader is used in the :ref:`http_application_update_sample` sample.

nRF5340 support (? should we include this, even though it isn't using multi-image builds yet - it still needs several images ?)
   nRF5340 contains two separate processors: a network core and an application core.

>    When programming applications to the nRF5340 PDK, they must be divided into two images, one for each core.

Reviewer: > they must be divided into at least one image per core.

Reviewer: unless you think it is obvious there may be multiple images per core.

Reviewer 3: Updated to:
 When programming applications to the nRF5340 PDK, they must be divided into at least two images, one for each core.

>    See :ref:`ug_nrf5340` for more information.

Reviewer: It's a bit premature, but I'm sure this will be valid before this is released, so I think it makes sense to include it.

   .. important::
      Currently, the two images must be built and programmed separately.

>       Multi-image builds are not yet supported for nRF5340.

Reviewer: @hakonfam : FYI, we need to remember to update these docs eventually.

##################

In many cases, the firmware that is programmed to a device consists of not only one application, but several separate images, where one of the images (the *parent image*) requires one or more other images (the *child images*) to be present.

> The child image then *chain-loads* (or *boots*) the parent image, which in turn might be a child image to another parent image and boot that one.

Reviewer: boots, not boot.

Reviewer 3: No - it's "which might boot ...".

When to use multiple images
***************************

An *image* (also referred to as executable, program, or elf file) consists of pieces of code and data that are identified by image-unique names recorded in a single symbol table.
The symbol table exists as metadata in a ``.elf`` or ``.exe`` file and is not included when the image is converted to a HEX file for programming.
Instead, the code and data are placed at addresses by a linker.
This linking process is what distinguishes images from object files (which do not require linking).
Therefore, to determine if you have zero, one, or more images, count the number of times the linker runs.

Using multiple images has the following advantages:

* You can run the linker multiple times and partition the final firmware into several regions.
  This partitioning is often useful for bootloaders.
* Since there is a symbol table for each image, the same symbol names can exist multiple times in the final firmware.
  This is useful for bootloader images, which might required their own copy of a library that the application uses, but in a different version or configuration.

> In the |NCS|, multiple images are required in the following scenarios:

Reviewer: Is it more correct to say "In NCS"?

I consider NCS to be a name, not a concept or even an abbreviation (In the Nordic Connect SDK does sound right).

So, to me it sounds like saying "In the Google", or "In the Facebook."

Reviewer 3: |NCS| is a shortcut for nRF Connect SDK, so this will resolve to "In the nRF Connect SDK".
We don't use the abbreviation NCS.

Reviewer: aha

  This is the default setting.
* Use a prebuilt HEX file of the child image and include it with the parent image.
* Ignore the child image.

When building the child image from source or using a prebuilt HEX file, the build system merges the HEX files of the parent and child image together so that they can easily be programmed in one single step.
This means that you might enable and integrate an additional image without even realizing it, just by using the default configuration.

To change the default configuration and configure how a child image is handled, locate the BUILD_STRATEGY configuration options for the respective child image in the parent image configuration.
For example, to use a prebuilt HEX file of the :ref:`secure_partition_manager` instead of building it, select :option:`CONFIG_SPM_BUILD_STRATEGY_USE_HEX_FILE` instead of the default :option:`CONFIG_SPM_BUILD_STRATEGY_FROM_SOURCE`, and specify the HEX file in :option:`CONFIG_SPM_HEX_FILE`.
To ignore an MCUboot child image, select :option:`CONFIG_MCUBOOT_BUILD_STRATEGY_SKIP_BUILD` instead of :option:`CONFIG_MCUBOOT_BUILD_STRATEGY_FROM_SOURCE`.

Defining and enabling a child image
***********************************

> You can enable existing child images in the |NCS| by enabling the respective modules in the parent image and selecting the desired build strategy. (? is that correct?)

Reviewer: This is correct

To turn an application that you have implemented into a child image that can be included in a parent image, you must add the required configuration options, update the build scripts to make it possible to enable the child image, and make sure that image-specific variables can be disambiguated.

Adding configuration options
============================

When enabling a child image, you select the *build strategy*, thus how the image is included.
The three options are:

* Build the child image from source along with the parent image- *IMAGENAME*\_BUILD_STRATEGY_FROM_SOURCE
* Merge the specified HEX file of the child image with the parent image - *IMAGENAME*\_BUILD_STRATEGY_USE_HEX_FILE, and *IMAGENAME*\_HEX_FILE to specify the HEX file
* Ignore the child image when building and build only the parent image - *IMAGENAME*\_BUILD_STRATEGY_SKIP_BUILD

You must add these four configuration options to the Kconfig file for your child image, replacing *IMAGENAME* with the (uppercase) name of your child image.

> (? are there any rules for the name? Does it have to be the same as the file name, for example?)

Reviewer: The name must match the "child image name" mentioned here:

https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1508/files#diff-99a904e387cfbc7dd8a3cc5e0cf2e7b3R138

Reviewer: It does not need to match any file name.

Reviewer 3: Updated to:
You must add these four configuration options to the Kconfig file for your child image, replacing *IMAGENAME* with the (uppercase) name of your child image (as specified in ``zephyr_add_child_image``).

Also, I moved the "Adding configuration options" section after the "Updating the build scripts" section so that it makes sense to refer to that function.

The following example shows the configuration options for MCUboot:

> (? this is mentioning zephyr_add_executable, which doesn't exist anymore?)

Reviewer: Should be renamed to 'zephyr_add_child_image'

Reviewer 3: Done. But - why zephyr_?

Reviewer: This is for legacy reasons, we used to believe this would be added upstream, which would require the zephyr_ prefix, now that it is not it is no longer necessary.

I'll remove the prefix so users don't think the function comes from upstream zephyr.

Reviewer: Prefix is removed

.. code-block:: Kconfig

   choice
  	prompt "MCUboot build strategy"
  	default MCUBOOT_BUILD_STRATEGY_FROM_SOURCE

   config MCUBOOT_BUILD_STRATEGY_USE_HEX_FILE
  	# Mandatory option when being built through 'zephyr_add_executable'
  	bool "Use HEX file instead of building MCUboot"

   if MCUBOOT_BUILD_STRATEGY_USE_HEX_FILE

   config MCUBOOT_HEX_FILE
  	# Mandatory option when being built through 'zephyr_add_executable'
  	string "MCUboot HEX file"

   endif # MCUBOOT_USE_HEX_FILE

   config MCUBOOT_BUILD_STRATEGY_SKIP_BUILD
  	# Mandatory option when being built through 'zephyr_add_executable'
  	bool "Skip building MCUboot"

   config MCUBOOT_BUILD_STRATEGY_FROM_SOURCE
  	# Mandatory option when being built through 'zephyr_add_executable'
  	bool "Build from source"

   endchoice

Updating the build scripts
==========================

To make it possible to enable a child image from a parent image, you must include the child image in the build script.
This code should be put in place in the cmake tree that is conditional on a configuration option for having the parent image use the child image.

> In the |NCS|, the code is included in the :file:`CMakeLists.txt` file for the samples. (? where is the mcuboot code?)

Reviewer: > (? where is the mcuboot code?)

Can you be more specific?

Reviewer 3: In samples/CMakeLists.txt, I see code like the one mentioned for SPM and for SECURE_BOOT. But not for mcuboot - where is that located?

Reviewer: It's here:

https://github.com/NordicPlayground/fw-nrfconnect-mcuboot/blob/master/zephyr/CMakeLists.txt#L86

.. code-block:: cmake

  set(child_image_name mcuboot)
  zephyr_add_executable(${child_image_name} require_build)

  set(build_directory ${CMAKE_CURRENT_BINARY_DIR}/mcuboot
  set(child_image_application_directory ${MCUBOOT_BASE}/boot/zephyr)

  if(${require_build})
    add_subdirectory(${child_image_application_directory} ${build_directory})
  endif()

> (? this code must be updated, and the following description as well)

Reviewer: Would you like me to do so?

Reviewer 3: Yes, please.

Defining and enabling a child image
***********************************

You can enable existing child images in the |NCS| by enabling the respective modules in the parent image and selecting the desired build strategy.
To turn an application that you have implemented into a child image that can be included in a parent image, you must update the build scripts to make it possible to enable the child image and add the required configuration options.
You should also know how image-specific variables are disambiguated and what targets of the child images are available.

Updating the build scripts
==========================

To make it possible to enable a child image from a parent image, you must include the child image in the build script.

This code should be put in place in the cmake tree that is conditional on a configuration option for having the parent image use the child image.
In the |NCS|, the code is included in the :file:`CMakeLists.txt` file for the samples, and in the MCUboot repository.

See the following example code:

> .. code-block:: cmake

Reviewer 3: I added the other two examples here since I feel they are more straight-forward - does that make sense or is it confusing?

Reviewer: That's ok.

.. code-block:: Kconfig

   choice
  	prompt "MCUboot build strategy"
  	default MCUBOOT_BUILD_STRATEGY_FROM_SOURCE

   config MCUBOOT_BUILD_STRATEGY_USE_HEX_FILE
  	# Mandatory option when being built through 'add_child_image'
  	bool "Use HEX file instead of building MCUboot"

   if MCUBOOT_BUILD_STRATEGY_USE_HEX_FILE

   config MCUBOOT_HEX_FILE
  	# Mandatory option when being built through 'add_child_image'
  	string "MCUboot HEX file"

   endif # MCUBOOT_USE_HEX_FILE

   config MCUBOOT_BUILD_STRATEGY_SKIP_BUILD
  	# Mandatory option when being built through 'add_child_image'
  	bool "Skip building MCUboot"

   config MCUBOOT_BUILD_STRATEGY_FROM_SOURCE
  	# Mandatory option when being built through 'add_child_image'
  	bool "Build from source"

   endchoice

Image-specific variables
========================

The child image and parent image are executed in different CMake processes and thus have different namespaces.
Variables in the parent image are not propagated to the child image, with the following exceptions:

* Any variable named *IMAGENAME*\_FOO in a parent image is propagated to the child image named *IMAGENAME* as FOO.

> * Variables that are in the list ``SHARED_MULTI_IMAGE_VARIABLES`` are propagated to all child images.

Reviewer 3: Where and how would you define that list?

Reviewer: It's a dynamic list, it could be amended from anywhere before add_child_image is invoked, including from CLI.

Currently only the variable EXTRA_KCONFIG_TARGETS and some variables related to it are in the list, but these variables are not interesting for 99% of users.

Not sure if/how to document this.

It is expected that eventually there will be more variables in this list, perhaps variables that users care about, at which point we should definitely list them here.

Reviewer 3: So basically, users would just define that list somewhere in their cmake file - correct?
That's what I would have guessed as well, so maybe we don't need to explain it in more detail.

Reviewer: Corret

With these two mechanisms, it is possible to set variables in child images from either parent images or the command line, and it is possible to set variables globally across all images.
For example, to change the ``CONF_FILE`` variable for the MCUboot image and the parent image, specify the CMake command as follows::

   cmake -Dmcuboot_CONF_FILE=prj_a.conf -DCONF_FILE=app_prj.conf

Child image targets
===================

You can indirectly invoke a selection of child image targets from the parent image.

> Currently, the child targets that can be invoked from the parent targets are ``menuconfig``, ``guiconfig``, and any targets listed in ``EXTRA_KCONFIG_TARGETS``.

Reviewer 3: Where is EXTRA_KCONFIG_TARGETS defined?

Reviewer: It is documented here: https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/kconfig.cmake#L51

and could be initialized from CLI. The primary user is SES, they use it to add their own Kconfig frontend.

Reviewer 3: OK, so probably not relevant for most users? Then let's leave as is.

To disambiguate targets, the same prefix convention is used as for variables.
This means that to run menuconfig, for example, you invoke the ``menuconfig`` target to configure the parent image and ``mcuboot_menuconfig`` to configure the MCUboot child image.

You can also invoke any child target directly from its build directory.

> Child build directories are located at the top of the parent's build directory.

Reviewer 3: What do you mean with the "top"? On root level in the build directory?

Reviewer: Yes

Reviewer 3: ok, changed

   endif # MCUBOOT_USE_HEX_FILE

   config MCUBOOT_BUILD_STRATEGY_SKIP_BUILD
  	# Mandatory option when being built through 'add_child_image'
  	bool "Skip building MCUboot"

   config MCUBOOT_BUILD_STRATEGY_FROM_SOURCE
  	# Mandatory option when being built through 'add_child_image'
  	bool "Build from source"

   endchoice

Image-specific variables
========================

The child image and parent image are executed in different CMake processes and thus have different namespaces.
Variables in the parent image are not propagated to the child image, with the following exceptions:

* Any variable named *IMAGENAME*\_FOO in a parent image is propagated to the child image named *IMAGENAME* as FOO.
* Variables that are in the list ``SHARED_MULTI_IMAGE_VARIABLES`` are propagated to all child images.

With these two mechanisms, it is possible to set variables in child images from either parent images or the command line, and it is possible to set variables globally across all images.
For example, to change the ``CONF_FILE`` variable for the MCUboot image and the parent image, specify the CMake command as follows::

   cmake -Dmcuboot_CONF_FILE=prj_a.conf -DCONF_FILE=app_prj.conf

> The command line that is used to create the child images can be extended by adding flags to the CMake variable ``EXTRA_MULTI_IMAGE_CMAKE_ARGS``.

Reviewer 3: ```suggestion
You can extend the build command that is used to create the child images by adding flags to the CMake variable ``EXTRA_MULTI_IMAGE_CMAKE_ARGS``.
```

Reviewer: No, invoking CMake is not building, it is configuring.

It's the ninja command that is the build command.

> This could for instance be used to get more debug information with the flag ``---trace-expand``.

Reviewer 3: ```suggestion
For example, add ``---trace-expand`` to that variable to output more debug information.
```

> Similarly the CMake variable ``EXTRA_MULTI_IMAGE_BUILD_OPT`` can be used to modify the options used when ninja is invoked on the child images.

Reviewer 3: ```suggestion
Similarly, you can modify the options for building the child image by adding them to the CMake variable ``EXTRA_MULTI_IMAGE_BUILD_OPT``.
```

Child image targets
===================

You can indirectly invoke a selection of child image targets from the parent image.
Currently, the child targets that can be invoked from the parent targets are ``menuconfig``, ``guiconfig``, and any targets listed in ``EXTRA_KCONFIG_TARGETS``.

To disambiguate targets, the same prefix convention is used as for variables.
This means that to run menuconfig, for example, you invoke the ``menuconfig`` target to configure the parent image and ``mcuboot_menuconfig`` to configure the MCUboot child image.

You can also invoke any child target directly from its build directory.

> Child build directories are located at the top of the parent's build directory.

Reviewer 3: ```suggestion
Child build directories are located at the root of the parent's build directory.
```

Reviewer 3: Think this change was lost somewhere ...

***********************************

You can enable existing child images in the |NCS| by enabling the respective modules in the parent image and selecting the desired build strategy.
To turn an application that you have implemented into a child image that can be included in a parent image, you must update the build scripts to make it possible to enable the child image and add the required configuration options.
You should also know how image-specific variables are disambiguated and what targets of the child images are available.

Updating the build scripts
==========================

To make it possible to enable a child image from a parent image, you must include the child image in the build script.

This code should be put in place in the cmake tree that is conditional on a configuration option for having the parent image use the child image.

> In the |NCS|, the code is included in the :file:`CMakeLists.txt` file for the samples, and in the MCUboot repository.

Reviewer: This is an internal implementation detail that they don't need to know.

Reviewer 3: Hmm, not sure I agree. When I'm reading this, I want to know what I need to do to make this work for my child image, and I want to see an example. Putting the code "somewhere in the tree" is very vague, and I think it helps to tell the user where they can look. So it's not meant as implementation detail, but as a pointer to where they can actually see it "in action".

Reviewer: I see, that's okay then.

   cmake -Dmcuboot_CONF_FILE=prj_a.conf -DCONF_FILE=app_prj.conf

The command line that is used to create the child images can be extended by adding flags to the CMake variable ``EXTRA_MULTI_IMAGE_CMAKE_ARGS``.
This could for instance be used to get more debug information with the flag ``---trace-expand``.

> Similarly the CMake variable ``EXTRA_MULTI_IMAGE_BUILD_OPT`` can be used to modify the options used when ninja is invoked on the child images.

Author: I would prefer to not use `ninja` here, but rather, something in line with: `the options used when the build command is invoked on the child images`.
as the user could be using `make` instead of ninja.

Although, I personally would prefer to inform the user of the `ninja -C <child_image_dir>` / `make -C <child_image_dir>` possibility, and then only introduce this extra argument if actually requested by users.

Reviewer: "ninja" is less ambiguous than "the build command" and there is precedence for directly using ninja in the docs.

Reviewer: https://docs.zephyrproject.org/latest/application/index.html

Author: but that page describes both `ninja` or `make`.
And states that `make` is an option on Linux and macOS, but for simplicity they will use `ninja` for rest of document.

Now, in this user guide we only use `ninja` and therefor I feel we should at least mention `make` as well.

@ru-fu Do you have an insight in the NCS doc on this. Do we in general only use `ninja` or do we also describe `make` as a possibility. 
To me, consistency is most important :)

Reviewer 3: In the NCS docs, we don't use ninja at all (except for the documentation build). We mention CMake in a few places, but in general we focus on SES. We are currently adding information about command line builds, but there we'll focus on west.
So I'm not in favor of mentioning ninja here.
That being said, I didn't get around to reviewing the addition earlier, so I have some more suggestions anyway. ;)

Reviewer 4: Is it possible to do the `ninja -C <sub_image> -v` option for `SES`? If not this option may still be needed.

Author: For SES right now, we might need an update.
Cause for parent image, SES is not invoking `ninja`, instead they parse the ninja files generated and invoke the compiler directly to have the output directly into SES.
For child images i'm unsure what will be done in long term (later updates to SES), but currently the build of sub-images are handled through CMake / ninja cause they are external projects, so there is kind of a mix of principles there, and SES know nothing about the sub-image.

So for SES, this could be a way to get verbose output, although user would have to go through
`Tools -> Options -> nRF Connect -> Additional CMake Options` and then re-import the project.
And then the same when disabling verbose output again.

Not even a very friendly solution in SES either.

Reviewer: @tejlmand : What if I change from "ninja" to "ninja or make" ?

I don't feel that this is necessary, but this appears to be subjective.

Reviewer 3: I added a suggestion further down that doesn't mention ninja or cmake. Maybe that'll do?

Reviewer: I don't think people know what a build command is, but since both you and @tejlmand support this phrasing I'll trust your judgement.

Reviewer: Will create another PR to address doc issues.

Reviewer: https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/1645
